rust/connlib/tunnel/src/tests/strategies.rs

```diff
@@ -107,33 +94,221 @@ pub(crate) fn tunnel_ip6s() -> impl Iterator<Item = Ipv6Addr> {
+                    (
+                        HashMap::<GatewayId, _>::default(),
+                        HashMap::<SiteId, HashSet<GatewayId>>::default(),
+                    ),
+                    |(mut gateways, mut sites), (gid, (gateway, site))| {
+                        sites.entry(site).or_default().insert(gid);
+                        gateways.insert(gid, gateway);
 
-            (gateways, portal)
-        })
+                        (gateways, sites)
+                    },
+                );
+
+                // For each DNS resource, we need to generate a set of DNS records.
+                let dns_resource_records = dns_resources
+                    .clone()
+                    .into_iter()
+                    .map(|resource| {
+                        let address = resource.address;
+
+                        match address.chars().next().unwrap() {
+                            '*' => subdomain_records(
+                                address.trim_start_matches("*.").to_owned(),
+                                domain_name(1..3),
+                            )
+                            .boxed(),
+                            '?' => subdomain_records(
+                                address.trim_start_matches("?.").to_owned(),
+                                domain_label(),
+                            )
+                            .boxed(),
+                            _ => resolved_ips()
+                                .prop_map(move |resolved_ips| {
+                                    HashMap::from([(address.parse().unwrap(), resolved_ips)])
+                                })
+                                .boxed(),
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .prop_map(|records| {
+                        let mut map = HashMap::default();
+
+                        for record in records {
+                            map.extend(record)
+                        }
+
+                        map
+                    });
+
+                let portal = StubPortal::new(
+                    gateways_by_site,
+                    gateway_selector,
+                    cidr_resources,
+                    dns_resources,
+                );
+
+                (Just(gateways), Just(portal), dns_resource_records)
+            },
+        )
+        .prop_filter(
+            "gateway must be able to access their assigned CIDR resources",
+            |(gateways, portal, _)| {
+                portal.cidr_resources().all(|(rid, r)| {
+                    portal
+                        .gateway_for_resource(*rid)
+                        .and_then(|g| gateways.get(g))
+                        .is_some_and(|g| {
+                            // TODO: PRODUCTION CODE DOES NOT HANDLE THIS!
+
+                            g.can_route_to(r.address)
+                        })
+                })
+            },
+        )
```

Comment:
This is a horribly inefficient filter right now and I'd like to get rid of it. The problem is that we currently only exercise the relay path "by accident" with gateways and clients that don't share an IP space (i.e. IPv4-only and IPv6-only). I'd like to make all gateways dual-stack in these tests so we can avoid this filter but for that, we first need to introduce some kind of `Firewall` that will purposely drop packets to force a connection through the relay.

Reply:
Fixed in https://github.com/firezone/firezone/pull/5917.